src/charm.py

```diff
@@ -13,91 +13,189 @@
-        # https://juju.is/docs/sdk/constructs#heading--statuses
+
+    def _manage_server_layer(self):
+        """Manage MLflow server layer w/ Pebble."""
+        container = self.unit.get_container("server")
+        mlflow_layer = self._mlflow_layer()
+        services = container.get_plan().to_dict().get("services", {})
+        if services.get("server") != mlflow_layer["services"]["server"]:
+            self.unit.status = MaintenanceStatus("MLflow server maintenance")
+            container.add_layer("mlflow-server", mlflow_layer, combine=True)
+            if container.get_service("server").is_running():
+                logging.info("Restarting MLflow server service")
+                container.stop("server")
+
+            container.start("server")
+
+    def _on_server_pebble_ready(self, event: PebbleReadyEvent):
+        """Define and start a workload using the Pebble API."""
+        # TODO: install mlflow in container or check if it's installed
+        self._manage_server_layer()
+
+        if not event.workload.get_service("server").is_running():
+            self.unit.status = BlockedStatus("The Mlflow server is not running.")
+            event.defer()
+            return
+
         self.unit.status = ActiveStatus()
 
     def _on_config_changed(self, _):
-        """Just an example to show how to deal with changed configuration.
-
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle config, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the config.py file.
-
-        Learn more about config at https://juju.is/docs/sdk/config
-        """
-        current = self.config["thing"]
-        if current not in self._stored.things:
-            logger.debug("found a new thing: %r", current)
-            self._stored.things.append(current)
-
-    def _on_fortune_action(self, event):
-        """Just an example to show how to receive actions.
-
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle actions, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the actions.py file.
-
-        Learn more about actions at https://juju.is/docs/sdk/actions
-        """
-        fail = event.params["fail"]
-        if fail:
-            event.fail(fail)
-        else:
-            event.set_results({"fortune": "A bug in the code is worth two in the documentation."})
+        """Handle the config-changed event."""
+        self._manage_server_layer()
+        self.ingress.update_config({
+            "service-hostname": self.config["host"], "service-port": self.config["port"]
+        })
+        self.unit.status = ActiveStatus()
+
+    def _dp_upgrade_action(self, event):
+        """Run MLflow dp upgrade."""
+        container = self.unit.get_container("server")
+        if event.params["i-really-mean-it"] is True:
```

Comment:
Sugestion: this could check `is False` and return, so the rest of lines would not need to be indented.